{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "cd599e5f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from pathlib import Path\n",
    "import numpy as np\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "f8bf8c9f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_sig(df, gene):\n",
    "    \"\"\"Return significance value or None.\"\"\"\n",
    "    vals = df.loc[df[\"Gene(s)\"] == gene, \"Significance\"]\n",
    "    if len(vals) == 0:\n",
    "        return None\n",
    "    val = vals.iloc[0]\n",
    "    return None if (isinstance(val, float) and np.isnan(val)) else val\n",
    "\n",
    "severity_rank = {\n",
    "    \"Pathogenic\": 6,\n",
    "    \"Likely pathogenic\": 5,\n",
    "    'Pathogenic/likely pathogenic/pathogenic, low penetrance': 4,\n",
    "    \"Risk factor\": 4,\n",
    "    \"Association\": 3,\n",
    "    \"Uncertain significance\": 2,\n",
    "    \"Uncertain risk allele\": 2,\n",
    "    \"Not provided\": 1,\n",
    "    None: 0,\n",
    "}\n",
    "\n",
    "def score_gene(info):\n",
    "    return max(\n",
    "        severity_rank.get(info[\"son\"], 0),\n",
    "        severity_rank.get(info[\"mother\"], 0),\n",
    "        severity_rank.get(info[\"father\"], 0),\n",
    "    )\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "4df6c66f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "len of overlapping genes: 32\n",
      "Unique significance categories: ['Likely pathogenic', None, 'Association', 'Not provided', 'Uncertain significance', 'Pathogenic', 'Risk factor', 'Pathogenic/likely pathogenic/pathogenic, low penetrance', 'Uncertain risk allele']\n",
      "\n",
      " Most significant overlapping genes:\n",
      "PERM1 {'son': 'Pathogenic', 'mother': 'Pathogenic', 'father': 'Pathogenic', 'score': 6}\n",
      "GBP1 {'son': 'Likely pathogenic', 'mother': 'Likely pathogenic', 'father': 'Likely pathogenic', 'score': 5}\n",
      "ECE1 {'son': 'Risk factor', 'mother': 'Risk factor', 'father': 'Risk factor', 'score': 4}\n",
      "ABCA4 {'son': None, 'mother': None, 'father': 'Pathogenic/likely pathogenic/pathogenic, low penetrance', 'score': 4}\n",
      "CHI3L1 {'son': 'Risk factor', 'mother': 'Risk factor', 'father': 'Risk factor', 'score': 4}\n",
      "FCGR2B {'son': 'Risk factor', 'mother': 'Risk factor', 'father': 'Risk factor', 'score': 4}\n",
      "MGST3 {'son': 'Association', 'mother': 'Association', 'father': 'Association', 'score': 3}\n",
      "BGLAP, PAQR6 {'son': 'Association', 'mother': 'Association', 'father': 'Association', 'score': 3}\n",
      "PADI2 {'son': 'Association', 'mother': 'Association', 'father': 'Association', 'score': 3}\n",
      "CYP4Z1, CYP4A22-AS1 {'son': 'Association', 'mother': 'Association', 'father': 'Association', 'score': 3}\n",
      "FCN3 {'son': 'Uncertain significance', 'mother': 'Uncertain significance', 'father': 'Uncertain significance', 'score': 2}\n",
      "PCSK9 {'son': None, 'mother': 'Uncertain significance', 'father': 'Uncertain significance', 'score': 2}\n",
      "SELENON {'son': None, 'mother': 'Uncertain significance', 'father': 'Uncertain significance', 'score': 2}\n",
      "ADAR {'son': 'Uncertain significance', 'mother': 'Uncertain significance', 'father': 'Uncertain significance', 'score': 2}\n",
      "TNFRSF1B {'son': 'Uncertain significance', 'mother': 'Uncertain significance', 'father': 'Uncertain significance', 'score': 2}\n",
      "APOA2 {'son': None, 'mother': 'Uncertain significance', 'father': 'Uncertain significance', 'score': 2}\n",
      "IL10, IL19 {'son': 'Uncertain risk allele', 'mother': 'Uncertain risk allele', 'father': 'Uncertain risk allele', 'score': 2}\n",
      "FCGR3A {'son': 'Uncertain significance', 'mother': 'Uncertain significance', 'father': 'Uncertain significance', 'score': 2}\n",
      "CRP {'son': 'Uncertain significance', 'mother': 'Uncertain significance', 'father': 'Uncertain significance', 'score': 2}\n",
      "NBPF1 {'son': 'Uncertain significance', 'mother': 'Uncertain significance', 'father': 'Uncertain significance', 'score': 2}\n",
      "RHCE {'son': 'Not provided', 'mother': 'Not provided', 'father': 'Not provided', 'score': 1}\n",
      "TNFRSF14 {'son': 'Not provided', 'mother': 'Not provided', 'father': 'Not provided', 'score': 1}\n",
      "TNFRSF14, TNFRSF14-AS1 {'son': 'Not provided', 'mother': 'Not provided', 'father': 'Not provided', 'score': 1}\n",
      "ZMPSTE24 {'son': 'Not provided', 'mother': 'Not provided', 'father': 'Not provided', 'score': 1}\n",
      "DISC1, TSNAX-DISC1 {'son': 'Not provided', 'mother': 'Not provided', 'father': 'Not provided', 'score': 1}\n",
      "TOR1AIP1 {'son': None, 'mother': None, 'father': None, 'score': 0}\n",
      "RYR2 {'son': None, 'mother': None, 'father': None, 'score': 0}\n",
      "CFH {'son': None, 'mother': None, 'father': None, 'score': 0}\n",
      "FLVCR1 {'son': None, 'mother': None, 'father': None, 'score': 0}\n",
      "NDUFS2 {'son': None, 'mother': None, 'father': None, 'score': 0}\n",
      "SPTA1 {'son': None, 'mother': None, 'father': None, 'score': 0}\n",
      "ITPKB {'son': None, 'mother': None, 'father': None, 'score': 0}\n"
     ]
    }
   ],
   "source": [
    "son_snp_df = pd.read_csv(Path(\"data/hg002-son-snp.csv\"), delimiter=\",\")\n",
    "father_snp_df = pd.read_csv(Path(\"data/hg003-father-snp.csv\"), delimiter=\",\")\n",
    "mother_snp_df = pd.read_csv(Path(\"data/hg004-mother-snp.csv\"), delimiter=\",\")\n",
    "\n",
    "son_genes = list(son_snp_df[\"Gene(s)\"])\n",
    "father_genes = list(father_snp_df[\"Gene(s)\"])\n",
    "mother_genes = list(mother_snp_df[\"Gene(s)\"])\n",
    "\n",
    "# overlap_gene_set = set(son_genes) & set(father_genes) & set(mother_genes)\n",
    "# print(\"len of overlapping genes\", len(overlap_gene_set))\n",
    "\n",
    "# sig_values = []\n",
    "# for gene in overlap_gene_set:\n",
    "#     significance = son_snp_df.loc[son_snp_df[\"Gene(s)\"] == gene, \"Significance\"].iloc[0]\n",
    "#     if significance not in sig_values:\n",
    "#         sig_values.append(significance)\n",
    "# print(sig_values)\n",
    "\n",
    "\n",
    "# compute overlap\n",
    "overlap_gene_set = set(son_genes) & set(father_genes) & set(mother_genes)\n",
    "print(\"len of overlapping genes:\", len(overlap_gene_set))\n",
    "\n",
    "overlap_info = {}\n",
    "sig_categories = []\n",
    "\n",
    "for gene in overlap_gene_set:\n",
    "    son_sig = get_sig(son_snp_df, gene)\n",
    "    mother_sig = get_sig(mother_snp_df, gene)\n",
    "    father_sig = get_sig(father_snp_df, gene)\n",
    "\n",
    "    overlap_info[gene] = {\n",
    "        \"son\": son_sig,\n",
    "        \"mother\": mother_sig,\n",
    "        \"father\": father_sig,\n",
    "    }\n",
    "\n",
    "    for val in (son_sig, mother_sig, father_sig):\n",
    "        if val not in sig_categories:\n",
    "            sig_categories.append(val)\n",
    "\n",
    "print(\"Unique significance categories:\", sig_categories)\n",
    "\n",
    "for gene, info in overlap_info.items():\n",
    "    overlap_info[gene][\"score\"] = score_gene(info)\n",
    "\n",
    "# sort genes\n",
    "ranked_genes = sorted(\n",
    "    overlap_info.items(),\n",
    "    key=lambda x: x[1][\"score\"],\n",
    "    reverse=True\n",
    ")\n",
    "\n",
    "# print top 10\n",
    "print(\"\\n Most significant overlapping genes:\")\n",
    "for g, info in ranked_genes:\n",
    "    print(g, info)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "58dc5075",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "hw",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
